### ISSUE TITLE
Gitches on Nintendo 64 with N64 Digital while using Controller/Rumble Pak / F-Zero

### ISSUE THREAD

**Issue by Taurus1099** [darthcloud/BlueRetro#264]
I’ve enabled the controller pak manager through the web interface and formatted it but a number of games prompt that there is a problem with the pak on launch. One example is Tony Hawk Pro Skater 1. As a separate issue, F-Zero does not detect the blueretro device as a controller. 

Console is modded with N64 digital and using latest version of the blueretro firmware. Unrelated, but with two NSO 64 controllers the number returned from the controller test tool for the right axis seems to be a bit higher than the benchmark. What setting in the advanced config should I adjust? Two different NSO 64 controllers return different values on the test tool, so there must be some variance at the factory with those sticks.

**Comment by darthcloud** [comment]
Try to format via a laptop if you used a phone, or just let tony hawk format it when prompted.

If detection fail you can simply press reset on console, also games give up too fast.

Play with the LX, RX maximum value in web config to fine tune joystick.

IMHO this is a waste of time, that test tool make people paranoid. Most games are programmed with a good play room.

**Comment by Taurus1100** [comment]
Thanks for replying! I've tried the suggestions but F-Zero does not work at all with a blueretro device and Tony Hawk will not recognize the controller pak. Testing a random sample of other controller pak games shows a major issue with the current integration. I've also experienced odd rumble behavior. On Goldeney 007, the rumble will not work correctly when firing. It gives a brief rumble and then stops when firing, when it should continue to rumble when firing.  This is a great product, and hopefully you'll be able to circle back around to the N64 to finish polishing the feature set. Thank you!

**Comment by darthcloud** [comment]
You might have some build related issues because I personally got none of those issues.

Sound like your data line is glitching somehow.

**Comment by Taurus1100** [comment]
This was confirmed from others using the adapter purchased from Tindie so it is likely the build. I’ll follow up with the seller.

**Comment by darthcloud** [comment]
If you can ask others to report feedback in this GitHub issues.

Everyone with issue should list the following information if possible:
* N64 motherboard revision
* List of internal mods installed
* Other external peripheral used

It could be problem with the dongle itself but it could also be some interoperability issue with some motherboard or some other mods.

**Comment by Taurus1100** [comment]
@bixxewoscht

**Comment by bixxewoscht** [comment]
I had mempak issues myself before, but they were resolved when I used a PC instead of a phone to change the config. These issues occured both on my adaptor as well as on an ESP32 Devkit which I used to compare functionality.
Regarding the other issues, both with a ESP32 Devkit as well as my own Adaptors I had sporadic rumble issues, close to the ones that you described. 
Nothing gamebreaking though, the core functionality always worked. To be honest I always accepted it as minor issues in a great open source project, maybe depending on different controllers, console revisions and mods used. In the issues section on GitHub there are also several topics related to these small issues, so I did not look more into it and accepted it as small software bugs that don't really affect me.

Now of course I don't want to blame you @darthcloud for anything in your amazing project and I want to offer the most for everyone who bought my Adaptor, so I will look further into it. 

Because I just moved last week, it will take more time than usual, I barely even have a TV set up at the moment, and my N64 consoles as well as the Adaptors are still in boxes.

I will try to get to it this weekend.

**Comment by darthcloud** [comment]
@bixxewoscht Anything is possible off course but the N64 is the original system BlueRetro was developed for and that I actually use myself a lot. I gone though GoldenEye a few time ever the last 3 years and never noticed any rumble issues. The way it is implemented makes the interface very robust as it is using the ESP RMT peripheral, (so no software bit bang). What you guys describe should not happen at all.

I'm using my own BlueRetro Devkit when playing  but I did test with a few clone ESP-DevKit-C in December and this spring and didn't got any controller pak error myself. Controller pak was tested by going through Turok 1 and Tony Hawk 1.

The issue sound like the data line is glitching.

My own board got ESD protection on each line using a TPD6F002DSVR chip. This got an equivalent resistance of 100ohm. No idea if this help or not.

Maybe chip me one of your dongle for me to test.

I play with early N64 rev mostly but got a few more now that I could try on too. I got no mod in my console.

**Comment by darthcloud** [comment]
@Taurus1100 / @Taurus1099 I got your adapter today and made some quick test and it look to be working fine for me.

See video:
https://youtu.be/IBnAhmHv9O8

I tested with your F-Zero X and TH1 cartridge with a NUS-CPU-04 and NUS-CPU-08-01 motherboards. Both work fine.

I got others N64 mobo in storage will dig them this weekend and report back.

Meanwhile can you confirm which mobo your N64 got? Did you install any mods? Are you using an original AC adapter?

**Comment by taurus1101** [comment]
Thanks for taking a look. My console revisions is NUS-CPU-08-01 (modded with N64 Digital) and I'm using the original power supply. I also tested with an additional power supply and no change. When you are able to do further testing, are you able to get into a game on F-Zero? I've seen the game respond to controller inputs for 10-15 seconds but eventually it stops completely. The N64 Digital menu is still accessible however. Also, any issues with the rumble in Goldeneye?

**Comment by darthcloud** [comment]
I think its likely the N64 digital that interferes with the pad data lines. I can test more this weekend but its likely all fine.

I don't know much about N64 digital but I think you could remove the one connection for pad and it would likely work fine. But you would lose the OSD access.

**Comment by darthcloud** [comment]
And what's up with using new account every comments?

**Comment by taurus1101** [comment]
Must be related to the N64 Digital. I noticed a similar issue on another open source project: 

https://github.com/Ryzee119/usb64/issues/42

**Comment by bixxewoscht** [comment]
I tested with my N64 with N64 Digital and had the same issue with F-Zero. Could not test with my stock N64 yet. But in my other tests I remember there were no issues with the game on a stock N64.

Thank you both for investigating further.

Holen Sie sich Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by darthcloud** [comment]
[n64_test_push_pull.zip](https://github.com/darthcloud/BlueRetro/files/8740052/n64_test_push_pull.zip)

This is untested but it change the data line from open-drain, to push-pull. Maybe that will help if the issue is on the TX side.

**Comment by taurus1101** [comment]
Everything is now working perfectly! The added ability to get vibration feedback on switching between rumble/memory pak is very helpful. Is there any functionality lost by changing the data line?

**Comment by astro187** [comment]
Hi @darthcloud, I'm the guy that slid into your DM's.  Similar to the above user, I can confirm that the new firmware fixes all of the issues I was having. I played with it for quite a while today and exercised all of the main functionality and everything worked great.  Thaks for finding a workaround so quickly... I was excited, then super bummed, and now excited again... this is a really neat implementation and will be my goto N64 setup going forward!
________________________________________________________________________________________________________________
Just in case it is useful info for future reference, when I _was_ having problems it was the following setup:
NUS-CPU-08-01 with N64Digital mod
stock PSU
Switch Pro controller
v1.5.1 of your firmware
@bixxewoscht receiver

What was happening (this was only for P1 on my modded system; P2 worked fine and my stock NUS-CPU-08-01 worked fine)
Controller pak emualation wasn't working at all (games either wouldn't recognize it or would report it corrupted)
Every few seconds an uncommanded controller input [Dpad  down + Dpad right + main joystick (1,-1)] would register

**Comment by mario-64** [comment]
I can confirm this firmware seems to resolve the issues I was having on my n64digital console as well, both with F-Zero X and overall memory pak issues.

**Comment by darthcloud** [comment]
I got an N64 digital this week and was finally able to investigate first hand.

I wasn't really a fan of the workaround I made to fix it, the N64 SI is means to have device with open drain IO. So switching to push-pull is not really ideal IMHO.

This is really an issue caused by the N64 Digital mode and I think the fix should be an HW on it.

To fix it I added a 74AHCT125N buffer between the the PIF pin 16 and the N64D ribbon COM pad

This did the trick F-Zero X menu doesn't glitch with BlueRetro and the N64D OSD still work as before.
![kicad_qEtttH5Odq](https://user-images.githubusercontent.com/3744056/184511049-a3c0653b-57f6-4340-9cb2-e3e6fdca5122.png)

The best would be to design a ribbon with a pad for the tiny 1 gate variant 74AHCT1G125W5-7

![PXL_20220813_205732992](https://user-images.githubusercontent.com/3744056/184511131-d7277b4b-5a69-4a79-9e85-c0df9101ace1.jpg)
![PXL_20220813_205811684](https://user-images.githubusercontent.com/3744056/184511135-e2215163-f73c-40cd-ad37-02035f4770d1.jpg)

**Comment by mario-64** [comment]
Will you please leave the software fix in place for those of us without skills to mod our systems? I would have to pay for this work to be done :-(

**Comment by darthcloud** [comment]
Problem is this fix doesn't work for everybody, and it actually break it for some people without a N64D.

Old FW stay available so you can keep what you currently have, and if any major update occur I can always post a special version in this thread

**Comment by mario-64** [comment]
Great thank you

**Comment by darthcloud** [comment]
Quick update PixelFX designed a small flex to fix this issue. I tested it and it work great.
![PXL_20220927_000325996](https://user-images.githubusercontent.com/3744056/194727468-1fed6d27-6be7-4753-922c-175e005fa6c1.jpg)

I'll update this issue once info on how to get it is available.

**Comment by timotchee** [comment]
@darthcloud Is this fix the same for both HW1 and HW2? ie for both external and internal (with detection) installs?

**Comment by darthcloud** [comment]
@timotchee Yes

**Comment by timotchee** [comment]
@darthcloud If I'm reading correctly, you prefer Dan's flex cable fix to your 74AHCT125N buffer method for electrical reasons?

**Comment by darthcloud** [comment]
They are both using 125 buffer, it's same thing except flex look nicer

**Comment by darthcloud** [comment]
Flex available here: https://www.pixelfx.co/n64digital-controller-flex-w-buffer.html

**Comment by BassXX** [comment]
Hello there, is there also a fix planned for UltraHDMI users? I updated to the latest firmware and now the system does not detect the controller at all anymore

**Comment by Sowden** [comment]
Yikes!  I have just ordered one of the dev kit chip to do an internal install on my UltraHDMI modded N64.  Has anyone reported a similar issue bundled with that mod?  If so, a new flex cable will have to be developed, and as nice as Marshal is, he's only one guy and has a busy life.  The chances of him developing one is slim, unless someone else steps in and develops one of their own.  But again, has anyone with a UltraHDMI mod had this problem?  Should this deveate into a different thread?

**Comment by pixelherogaming** [comment]
> Flex available here: https://www.pixelfx.co/n64digital-controller-flex-w-buffer.html

This flex doesn't seem available anymore?

I am having an issue with my N64 digital modded system along with the blue retro external mod. My NSO and 64 controller will sync. And then randomly will disconnect mid game. Any work around?

**Comment by timotchee** [comment]
Dan of pixelfx/blackdogtech said there will be more flex cables in the future, but without a date.  I would guess in the next 5 months.  Or you can build your own circuit w/buffer with the info above.  If you go the FFC route, you will need to solder surface mount components onto the FFC.  Below I've listed the component to DIY, and the SMD components Dan originally specified for his flex cables, but he removed the FFC web pages so it's possible they change in the future.  There's no difference between the two, the FFC simply looks nicer if you're into staring at the innards of your console.  The DIY option will be easier to solder.

Buffer for DIY circuit - TI SN74AHCT125N: https://mou.sr/3ivItle or https://www.digikey.com/short/9z5dz7pp
FFC N64 Buffer - Nexperia 74AHC1G17GVH: https://mou.sr/3gPJYKN or https://www.digikey.com/short/p7jhvrff
FFC N64/PS1 Ceramic Capacitor(s) - 0.1uF 10V 0402: https://mou.sr/3uhdFaF or https://www.digikey.com/short/qt4nq3bq
FFC PS1 Buffer - TI SN74HCS125QPWRQ1: https://mou.sr/3H6pdoF or https://www.digikey.com/short/0mq77pbf

Please double check above and the FFC listing whenever it returns to ensure I made no mistakes.

**Comment by pixelherogaming** [comment]
> Dan of pixelfx/blackdogtech said there will be more flex cables in the future, but without a date. I would guess in the next 5 months. Or you can build your own circuit w/buffer with the info above. If you go the FFC route, you will need to solder surface mount components onto the FFC. Below I've listed the component to DIY, and the SMD components Dan originally specified for his flex cables, but he removed the FFC web pages so it's possible they change in the future. There's no difference between the two, the FFC simply looks nicer if you're into staring at the innards of your console. The DIY option will be easier to solder.
> 
> Buffer for DIY circuit - TI SN74AHCT125N: https://mou.sr/3ivItle or https://www.digikey.com/short/9z5dz7pp FFC N64 Buffer - Nexperia 74AHC1G17GVH: https://mou.sr/3gPJYKN or https://www.digikey.com/short/p7jhvrff FFC N64/PS1 Ceramic Capacitor(s) - 0.1uF 10V 0402: https://mou.sr/3uhdFaF or https://www.digikey.com/short/qt4nq3bq FFC PS1 Buffer - TI SN74HCS125QPWRQ1: https://mou.sr/3H6pdoF or https://www.digikey.com/short/0mq77pbf
> 
> Please double check above and the FFC listing whenever it returns to ensure I made no mistakes.

So I should just need to order the buffer for diy (since i dont care what it looks like) and them just solder like example above? Or do I also need capacitor(s)

**Comment by timotchee** [comment]
Yeah, DIY is easiest and fastest.  I don't see any caps in the circuit, but @darthcloud is the expert, if he sees this.

**Comment by mario-64** [comment]
My recommendation is to avoid firmware 1.8 and above or, alternatively, stop using BlueRetro for N64. We shouldn’t have to further mod our consoles.

**Comment by L10N37** [comment]
I had a customer send me in the flex kit and his N64 that I put the HDMI kit in. This flex seems to kill the ability for a normal N64 controller to enter the OSD?

**Comment by darthcloud** [comment]
@L10N37 It shouldn't do that, probably problem with flex install.

**Comment by L10N37** [comment]
ah it was like 3/ 4am and pulled 2 all nighters in a row, i just missed the 5v connection

**Comment by darthcloud** [comment]
For those that will not fix their N64 Digital install and that liked the imperfect workaround here a build of v1.8 that reintegrate it.

[v1.8-push_pull_port1_hw1_n64.zip](https://github.com/darthcloud/BlueRetro/files/10239311/v1.8-push_pull_port1_hw1_n64.zip)

**Comment by pixelherogaming** [comment]
> For those that will not fix their N64 Digital install and that liked the imperfect workaround here a build of v1.8 that reintegrate it.
> 
> [v1.8-push_pull_port1_hw1_n64.zip](https://github.com/darthcloud/BlueRetro/files/10239311/v1.8-push_pull_port1_hw1_n64.zip)

Any latency or disadvantage to this work around as opposed to adding a buffer?

**Comment by mario-64** [comment]
> For those that will not fix their N64 Digital install and that liked the imperfect workaround here a build of v1.8 that reintegrate it.
> 
> [v1.8-push_pull_port1_hw1_n64.zip](https://github.com/darthcloud/BlueRetro/files/10239311/v1.8-push_pull_port1_hw1_n64.zip)

Thank you!

**Comment by mario-64** [comment]
The fix is integrated in firmwares prior to 1.8 but it was removed in the main branch.

**Comment by darthcloud** [comment]
Its a hardware bug and the workaround only work for half people with n64 digital, and break half people without one.

The real fix is the pixelfx patch cable.

**Comment by mario-64** [comment]
> Its a hardware bug and the workaround only work for half people with n64 digital, and break half people without one.
> 
> The real fix is the pixelfx patch cable.

Blueretro, N64 or n64digital hardware bug?

**Comment by darthcloud** [comment]
n64 digital hardware bug, this is why they provide a flex patch

**Comment by BassXX** [comment]
As an ultra hdmi user, I suffer the same problem. 

And I know for sure that there won’t be a fixed flex cable, at least not from the mod creator.

so basically I have no choice but to either switch to N64Digital or use those workarounds. I hope those will be provided in the future versions as well. Thank you

**Comment by bixxewoscht** [comment]
You could also look at the Flex/DIY fix documentation and adapt it for Ultra HDMI. There shouldn't be too much of a difference. But I don't know it exactly since I've never looked at Ultra HDMI.

**Comment by darthcloud** [comment]
You just need to do as I explained in my previous post in this thread
https://github.com/darthcloud/BlueRetro/issues/264#issuecomment-1214224340

**Comment by Sowden** [comment]
Would the #264 post fix for those with UltraHDMI's as well?

**Comment by Sowden** [comment]
@darthcloud Ok, let me see if I get this one strait.  The only wire you are intermediating is the N64D ribbon cable to the PIF 16.  You aren't messing with any of the other connections?  I'm asking because I'm finding the similarities between the N64D and the UltraHDMI.  The UHDMI has a ribbon cable that I believe you attach to the same spots as the CON connection on the N64D.  

![ezgif com-gif-maker](https://user-images.githubusercontent.com/3219909/208580080-6b274a59-5234-409e-bd3d-e01059c4f106.jpg)
(not my photo, taken to show example)
So would that mean that you would connect the UHDMI ribbon -> Pin 2 <74AHCT125N> Pin 3 -> (whatever pin depending on your motherboard)?  Cause that might be the solution for the select few that own UltraHDMI sets.

**Comment by darthcloud** [comment]
@Sowden Yeah just piggy-back a 74AHCT125N on top of the PIF chip like I did (connect 3V3 & GND to pif pin directly) then the pin label N64D_COM (pin3 ) is the one you plug your ultra HDMI cable to. and pin 2 goes to the pif controller 1 pin. Don't forget to GND all the needed pins as done in my schematic.

**Comment by Sowden** [comment]
Oh, GND Pin 1 -> PIF 14 and 3.3 v Pin 14 -> PIF 28?  So the 74AHCT125N has four connections in total, is that correct?

**Comment by nortakales** [comment]
Just wanted to pop in here and see if my issue might be related. I've got an N64 with the [Gamebox Systems 64HD](https://castlemaniagames.com/products/gamebox-systems-64hd-digital-video-output-kit). When I try to use my [BlueRetro adapter (from Retrotime)](https://8bitmods.com/n64-blueretro-bt-controller-receiver-with-memory-pak-original-grey/) in controller slot 1, I am getting graphical glitches only on my HDMI out (regular composite output is unaffected). Other controller slots do not have the issue. It's almost as if (and this is just a guess) the OSD menu is being interacted with even though I haven't brought it up. I haven't tried your firmware posted here yet though, I'll give it a try tomorrow.

**Comment by bixxewoscht** [comment]
From what I heard from other users (I don't have a 64HD) the Gamebox 64HD definitely has the same issue as N64Digital and other mods that tap into the controller lines. Unfortunately the known fix was not implemented from the get go on the N64HD.

**Comment by nortakales** [comment]
I updated my device with [v1.8-push_pull_port1_hw1_n64.zip](https://github.com/darthcloud/BlueRetro/files/10239311/v1.8-push_pull_port1_hw1_n64.zip) and it did not resolve the graphical glitches :(

**Comment by nortakales** [comment]
To be clear, with the 64HD I am NOT getting issues with F-Zero or Tony Hawk as reported here for N64 Digital and Ultra HDMI mods.

I am just getting graphical glitches when my BlueRetro adapter is in controller port one. The glitches usually look like a bunch of random colors flash across the bottom half of the screen. The game I have found it most noticeable on is Snowboard Kids. Here is a video: https://photos.app.goo.gl/kAvNYDFvx1Ush2bP8

**Comment by bixxewoscht** [comment]
I heard from people with a N64 HD that it has problems with signal integrity as it is, without using blueretro simultaneously. Reason is most likely the flex-pcb design thats not up to industry standards (ground return path). So likely when using blueretro this amplifies the issue.

**Comment by nortakales** [comment]
Yea, I have noticed very infrequently some visual issues, but mostly it is fine. BlueRetro is increasing the frequency by like 100x or 1000x though.

**Comment by s-taylor** [comment]
I've ordered the flex cable, I am having issues where the OSD is occasionally inaccessible. Where can I find the installation instructions? I have a friend who is experienced at soldering, hope he can help.

**Comment by Z3R0XZ** [comment]
I get graphical glitches, video desync, and signal loss from my N64 Digital via HDMI using Blue Retro. Is this problem related?

**Comment by browntownington** [comment]
Anyone know if this issue was resolved with the Retro Gem? I would assume it was  but haven't read anything to confirm.

**Comment by mario-64** [comment]
Doubt it. N64 RetroGem = N64 digital. Same device just different branding.

**Comment by browntownington** [comment]
From what I saw in this thread it was confirmed issue was a n64 digital hardware bug. Pixelfx were selling a flex cable specifically for n64digtal, but i see no mention of it for the retrogem. 
As for N64 RetroGem = N64 digital. They use different firmware. Also the hardware design and components are not exactly the same. Pixelfx were aware of this bug. Hence why I'm curious if it was taken into consideration. I'll ask on their discord.

**Comment by teconmoon** [comment]
> For those that will not fix their N64 Digital install and that liked the imperfect workaround here a build of v1.8 that reintegrate it.
> 
> [v1.8-push_pull_port1_hw1_n64.zip](https://github.com/darthcloud/BlueRetro/files/10239311/v1.8-push_pull_port1_hw1_n64.zip)

@darthcloud I know this is a workaround and not a real fix, but any chance of getting a firmware 1.9.1 with the same modification to change the data line from open-drain to push-pull? I was hoping to use my 8BitDo Mod Kit controller with my BlueRetro HW1 adapter, and there were some related fixes in 1.9 that I was hoping to take advantage of.

I have an N64Digital HW1, and besides not being able to find the flex cable that fixes the core problem, my soldering skills are horrendous and I'd rather not pay more to have someone do the installation (the original N64Digital mod and install was already pretty expensive).

I understand if this use-case isn't something you can keep supporting. Thanks!

**Comment by ottelo9** [comment]
This is my quick'n'dirty implementation:
![PIF N64Digital fix ](https://github.com/user-attachments/assets/b9183507-fad7-4769-aa48-b2f447ebfb1f)
![IMG_5688](https://github.com/user-attachments/assets/9bfa54fa-beb8-4bbb-824b-f82350e632e2)

**Comment by Kitocco** [comment]
For anyone else with an N64Digital, I found the BlueRetro flex still available on PixelFX's site. It's just under the legacy replacement parts page here: https://www.pixelfx.co/product-page/replacement-parts 

I haven't bought it myself and honestly, I'm unsure if I will since I know I'll have to send it again for an install. I got mine installed back in Summer 2022 by Rainy Day Retro (wonderful mod shop, highly recommend), but at this point... I think I may just resort to playing Controller Pak games on my Mister Pi.
And I don't trust myself with installing it after a botched GBA SP USB-c mod install in which I ripped some pads in pursuit of better alignment. and the high temps needed for lead-free solder.
